src/gui/src/clockWidget.h

```diff
@@ -0,0 +1,398 @@
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <QComboBox>
+#include <QDockWidget>
+#include <QGraphicsItem>
+#include <QGraphicsScene>
+#include <QGraphicsView>
+#include <QMenu>
+#include <QPainterPath>
+#include <QPushButton>
+#include <QTabWidget>
+#include <QToolTip>
+#include <variant>
+
+#include "gui/gui.h"
+#include "staGuiInterface.h"
+
+namespace sta {
+class dbSta;
+}
+
+namespace utl {
+class Logger;
+}
+
+namespace gui {
+
+class ClockNodeGraphicsViewItem;
+
+class ClockTreeRenderer : public Renderer
+{
+ public:
+  ClockTreeRenderer(ClockTree* tree);
+  ~ClockTreeRenderer() {}
+
+  virtual void drawObjects(Painter& painter) override;
+
+  void setPathTo(odb::dbITerm* term);
+  void clearPathTo();
+
+ private:
+  ClockTree* tree_;
+
+  odb::dbITerm* path_to_;
+};
+
+class ClockNetGraphicsViewItem : public QGraphicsItem
+{
+ public:
+  ClockNetGraphicsViewItem(
+      odb::dbNet* net,
+      const std::vector<ClockNodeGraphicsViewItem*>& source_nodes,
+      const std::vector<ClockNodeGraphicsViewItem*>& sink_nodes,
+      QGraphicsItem* parent = nullptr);
+
+  QRectF boundingRect() const override;
+  void paint(QPainter* painter,
+             const QStyleOptionGraphicsItem* option,
+             QWidget* widget) override;
+
+  virtual QPainterPath shape() const override { return path_; }
+
+  void buildPath();
+
+ private:
+  odb::dbNet* net_;
+  std::vector<ClockNodeGraphicsViewItem*> source_nodes_;
+  std::vector<ClockNodeGraphicsViewItem*> sink_nodes_;
+
+  QPainterPath path_;
+
+  constexpr static qreal trunk_rounding_ = 0.1;
+  constexpr static qreal leaf_rounding_ = 0.1;
+
+  void buildTrunkJunction(const std::vector<ClockNodeGraphicsViewItem*>& nodes,
+                          bool top_anchor,
+                          const QPointF& trunk_junction);
+  void addLeafPath(const QPointF& start,
+                   const qreal y_trunk,
+                   const QPointF& end);
+
+  void setNetInformation();
+};
+
+class ClockNodeGraphicsViewItem : public QGraphicsItem
```

Comment:
A class level comment on the *ViewItem classes would be helpful.  Its a bit hard to guess from their name what they are.

Reply:
added some comments